### PR TITLE
Issue#110 add telecom conversion

### PR DESCRIPTION
**Pull request by cragun47**
This adds Telecom conversion and associated unit tests.

See Issue for description of conversion.

### CODE REVIEW

File: src/test/java/io/github/linuxforhealth/hl7/data/Hl7RelatedGeneralUtilsTest.java [hunk 1]
```diff
@@ -177,5 +177,35 @@ public void test_getAddressDistrict() {
...
+    assertThat(Hl7RelatedGeneralUtils.getFormattedTelecomNumberValue("111","22","333","4444444","x555","112")).contains("4444");
+    assertThat(Hl7RelatedGeneralUtils.getFormattedTelecomNumberValue("111","22","333","4444444","x555","112")).contains("x555");
+    assertThat(Hl7RelatedGeneralUtils.getFormattedTelecomNumberValue("111","22","333","4444444","x555","112")).isEqualTo("+22 333 444 4444 x555");
+        // Country, Area, and Extension are used if there is a Local number, and XTN1 and XTN12 are ignored  
+    assertThat(Hl7RelatedGeneralUtils.getFormattedTelecomNumberValue("111","","333","4444444","x555","112")).isNotEmpty();
+    assertThat(Hl7RelatedGeneralUtils.getFormattedTelecomNumberValue("111","","333","4444444","x555","112")).contains("333");
+    assertThat(Hl7RelatedGeneralUtils.getFormattedTelecomNumberValue("111","","333","4444444","x555","112")).contains("4444");
+    assertThat(Hl7RelatedGeneralUtils.getFormattedTelecomNumberValue("111","","333","4444444","x555","112")).contains("x555");
+    assertThat(Hl7RelatedGeneralUtils.getFormattedTelecomNumberValue("111","","333","4444444","x555","112")).isEqualTo("(333) 444 4444 x555");
```
Comment (LisaWellman) [inline src/test/java/io/github/linuxforhealth/hl7/data/Hl7RelatedGeneralUtilsTest.java:207]: Suggest that we add some tests:
- extension not included
- country code not included
- area code not included

I am not sure how international numbers would come in, maybe they would be unformatted - we can add more later when we see real examples.

File: src/test/java/io/github/linuxforhealth/hl7/segments/Hl7TelecomFHIRConversionTest.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    ;
+
+    Patient patient = PatientUtils.createPatientFromHl7Segment(patientPhone);
+    assertThat(patient.hasTelecom()).isTrue();
+    List<ContactPoint> contacts = patient.getTelecom();
+    assertThat(contacts.size()).isEqualTo(3);
+
+    // First home contact
+    ContactPoint contact = contacts.get(0); 
+    assertThat(contact.getUse()).hasToString("HOME");
```
Comment (LisaWellman) [inline src/test/java/io/github/linuxforhealth/hl7/segments/Hl7TelecomFHIRConversionTest.java:41]: The value of both use and system should be lower case.  I am not sure if these methods are checking case, but we should lowercase to be most accurate even if they do not.

File: src/test/java/io/github/linuxforhealth/hl7/segments/Hl7TelecomFHIRConversionTest.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    assertThat(contact.getUse()).hasToString("MOBILE");
+    assertThat(contact.getValue()).hasToString("+22 555 222 1313");
+    assertThat(contact.hasRank()).isTrue();
+    assertThat(contact.getRank()).hasToString("1");
+    assertThat(contact.getSystem()).hasToString("PHONE");
+
+    // First work contact
+    contact = contacts.get(2); 
+    assertThat(contact.getUse()).hasToString("WORK");
+    assertThat(contact.getValue()).hasToString("(555) 111 1414 x889");
```
Comment (LisaWellman) [inline src/test/java/io/github/linuxforhealth/hl7/segments/Hl7TelecomFHIRConversionTest.java:58]: as mentioned elsewhere, should be `ext 889`

File: src/test/java/io/github/linuxforhealth/hl7/segments/Hl7TelecomFHIRConversionTest.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+  public ExpectedException exceptionRule = ExpectedException.none();
+
+
+  @Test
+
+  public void patient_telcom_test() {
+
+    String patientPhone =
+    "MSH|^~\\&|MIICEHRApplication|MIIC|MIIC|MIIC|201705130822||VXU^V04^VXU_V04|test1100|P|2.5.1|||AL|AL|||||Z22^CDCPHINVS|^^^^^MIIC^SR^^^MIIC|MIIC\n"
+    + "PID|1||12345678^^^^MR|ALTID|Moose^Mickey^J^III^^^||20060504|M|||||^PRN^PH^^22^555^1111313^^^^^^^^^^^2~^PRN^CP^^22^555^2221313^^^^^^^^^^^1|^PRN^PH^^^555^1111414^x889||||||||||||||||\n"
```
Comment (LisaWellman) [inline src/test/java/io/github/linuxforhealth/hl7/segments/Hl7TelecomFHIRConversionTest.java:31]: As mentnioned previously, I don't think x889 is a valid extension as it is numeric

**Review by LisaWellman — CHANGES_REQUESTED** [review]
A few comments, mainly want to address extension

File: src/main/java/io/github/linuxforhealth/hl7/data/Hl7RelatedGeneralUtils.java [hunk 5]
```diff
@@ -239,4 +239,39 @@ public static String getCodeableConceptDisplay(CodeableConcept codeableConcept)
         return codeableConcept.getCodingFirstRep().getDisplay();
     }
 
+    // Takes all the pieces of telecom number from XTN, formats to a user friendly Telecom number based on rules documented in the steps
+    public static String getFormattedTelecomNumberValue(String xtn1Old, String xtn5Country, String xtn6Area, String xtn7Local, String xtn8Extension, String xtn12Unformatted) {
+        String returnValue = "";
+
+        // If the local number exists, concatenate country, area, local.
+        if (xtn7Local != null && xtn7Local.length() > 0) {
```
[the remark this answers is not in the document]
Comment (cragun47) [reply src/main/java/io/github/linuxforhealth/hl7/data/Hl7RelatedGeneralUtils.java:247]: Agree.  Will do.

File: src/main/java/io/github/linuxforhealth/hl7/data/Hl7RelatedGeneralUtils.java [hunk 6]
```diff
@@ -239,4 +239,39 @@ public static String getCodeableConceptDisplay(CodeableConcept codeableConcept)
...
+            if (xtn5Country != null && xtn5Country.length() > 0) {
+                // If there is a country code, format +22 123 456 7890
+                returnValue = "+" + xtn5Country + " " + xtn6Area + " " + formatLocalNumber(xtn7Local);
+            } else {
+                // If there is a NOT a country code, format (123) 456 7890
+                returnValue = "(" + xtn6Area + ") " + formatLocalNumber(xtn7Local);
+            }
+            // If an exrention exists, append a space and the extension
```
[the remark this answers is not in the document]
Comment (cragun47) [reply src/main/java/io/github/linuxforhealth/hl7/data/Hl7RelatedGeneralUtils.java:255]: OK

File: src/test/java/io/github/linuxforhealth/hl7/segments/Hl7TelecomFHIRConversionTest.java (hunk 2 above)
Comment (cragun47) [reply src/test/java/io/github/linuxforhealth/hl7/segments/Hl7TelecomFHIRConversionTest.java:41]: The values from the .toString are coming back caps:

>contact.getUse().toString()
"HOME"
contact.getUse()
ContactPoint$ContactPointUse@166 "HOME"

I think a better way is to use the ENUM constants of the class.  I have changed the tests similar to this:

`assertThat(contact.getUse()).isEqualTo(ContactPoint.ContactPointUse.HOME);`

File: src/test/java/io/github/linuxforhealth/hl7/segments/Hl7TelecomFHIRConversionTest.java (hunk 3 above)
Comment (cragun47) [reply src/test/java/io/github/linuxforhealth/hl7/segments/Hl7TelecomFHIRConversionTest.java:58]: To match the standard you cite, I have changed it numeric content with ` ext. ` as prefix, including surrounding spaces.

File: src/main/resources/hl7/datatype/Telecom.yml [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+          xtn5Country: String, XTN.5
+          xtn6Area: String, XTN.6
+          xtn7Local: String, XTN.7
+          xtn8Extension: String, XTN.8
+          xtn12Unformatted: String, XTN.12
+
+# Default use to "home" or "work", whatever is passed in
+use_1:
+     type: STRING
+     valueOf: $use
```
[the remark this answers is not in the document]
Comment (cragun47) [reply src/main/resources/hl7/datatype/Telecom.yml:29]: Use is passed in from the Patient.yml.  Remember there are two phone sections, so they will each have an initial use:
```
telecom_1:
    valueOf: datatype/Telecom
    generateList: true
    expressionType: resource
    specs: PID.14
    constants: 
       use: "work"

         
telecom_2:
    valueOf: datatype/Telecom
    generateList: true
    expressionType: resource
    specs: PID.13
    constants: 
       use: "home"
```

File: src/test/java/io/github/linuxforhealth/hl7/data/Hl7RelatedGeneralUtilsTest.java [hunk 8]
```diff
@@ -177,5 +177,35 @@ public void test_getAddressDistrict() {
...
+    assertThat(Hl7RelatedGeneralUtils.getFormattedTelecomNumberValue("","22","333","","x555","")).isEmpty();
+    // XTN12 and XTN1 will be used if no local number
+    assertThat(Hl7RelatedGeneralUtils.getFormattedTelecomNumberValue("111","22","333","","x555","")).isEqualTo("111");
+    assertThat(Hl7RelatedGeneralUtils.getFormattedTelecomNumberValue("","22","333","","x555","112")).isEqualTo("112");
+    // Country, Area, and Extension are used if there is a Local number, and XTN1 and XTN12 are ignored  
+    assertThat(Hl7RelatedGeneralUtils.getFormattedTelecomNumberValue("111","22","333","4444444","x555","112")).isNotEmpty();
+    assertThat(Hl7RelatedGeneralUtils.getFormattedTelecomNumberValue("111","22","333","4444444","x555","112")).contains("+22");
+    assertThat(Hl7RelatedGeneralUtils.getFormattedTelecomNumberValue("111","22","333","4444444","x555","112")).contains("333");
+    assertThat(Hl7RelatedGeneralUtils.getFormattedTelecomNumberValue("111","22","333","4444444","x555","112")).contains("4444");
+    assertThat(Hl7RelatedGeneralUtils.getFormattedTelecomNumberValue("111","22","333","4444444","x555","112")).contains("x555");
```
[the remark this answers is not in the document]
Comment (cragun47) [reply src/test/java/io/github/linuxforhealth/hl7/data/Hl7RelatedGeneralUtilsTest.java:200]: Done

File: src/test/java/io/github/linuxforhealth/hl7/data/Hl7RelatedGeneralUtilsTest.java [hunk 9]
```diff
@@ -177,5 +177,35 @@ public void test_getAddressDistrict() {
...
+    // Everything empty except XTN1 returns XTN1.
+    assertThat(Hl7RelatedGeneralUtils.getFormattedTelecomNumberValue("111","","","","","")).isEqualTo("111");
+    // Everything empty except XTN12 returns XTN12.
+    assertThat(Hl7RelatedGeneralUtils.getFormattedTelecomNumberValue("","","","","","112")).isEqualTo("112");
+    // XTN12 takes priority over XTN1
+    assertThat(Hl7RelatedGeneralUtils.getFormattedTelecomNumberValue("111","","","","","112")).isEqualTo("112");
+    // Country, Area, and Extension are ignored if there is no Local number
+    assertThat(Hl7RelatedGeneralUtils.getFormattedTelecomNumberValue("","22","333","","x555","")).isEmpty();
+    // XTN12 and XTN1 will be used if no local number
+    assertThat(Hl7RelatedGeneralUtils.getFormattedTelecomNumberValue("111","22","333","","x555","")).isEqualTo("111");
```
[the remark this answers is not in the document]
Comment (cragun47) [reply src/test/java/io/github/linuxforhealth/hl7/data/Hl7RelatedGeneralUtilsTest.java:193]: Fixed

File: src/main/java/io/github/linuxforhealth/hl7/data/Hl7RelatedGeneralUtils.java [hunk 10]
```diff
@@ -239,4 +239,39 @@ public static String getCodeableConceptDisplay(CodeableConcept codeableConcept)
...
+            if (xtn5Country != null && xtn5Country.length() > 0) {
+                // If there is a country code, format +22 123 456 7890
+                returnValue = "+" + xtn5Country + " " + xtn6Area + " " + formatLocalNumber(xtn7Local);
+            } else {
+                // If there is a NOT a country code, format (123) 456 7890
+                returnValue = "(" + xtn6Area + ") " + formatLocalNumber(xtn7Local);
+            }
+            // If an exrention exists, append a space and the extension 
+            if (xtn8Extension != null && xtn8Extension.length() > 0) {
+                returnValue = returnValue + " " + xtn8Extension;
```
[the remark this answers is not in the document]
Comment (cragun47) [reply src/main/java/io/github/linuxforhealth/hl7/data/Hl7RelatedGeneralUtils.java:257]: Changed to " ext. " to match the detail in the standard.

File: src/main/java/io/github/linuxforhealth/hl7/data/Hl7RelatedGeneralUtils.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (cragun47) [reply src/main/java/io/github/linuxforhealth/hl7/data/Hl7RelatedGeneralUtils.java:255]: Fixed

File: src/main/java/io/github/linuxforhealth/hl7/data/Hl7RelatedGeneralUtils.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (cragun47) [reply src/main/java/io/github/linuxforhealth/hl7/data/Hl7RelatedGeneralUtils.java:247]: Fixed.  Added new logic and refactored to appease the linter.

File: src/test/java/io/github/linuxforhealth/hl7/data/Hl7RelatedGeneralUtilsTest.java (hunk 1 above)
Comment (cragun47) [reply src/test/java/io/github/linuxforhealth/hl7/data/Hl7RelatedGeneralUtilsTest.java:207]: Added additional tests as suggested.

File: src/test/java/io/github/linuxforhealth/hl7/segments/Hl7TelecomFHIRConversionTest.java (hunk 4 above)
Comment (cragun47) [reply src/test/java/io/github/linuxforhealth/hl7/segments/Hl7TelecomFHIRConversionTest.java:31]: Fixed.

File: src/test/java/io/github/linuxforhealth/hl7/data/Hl7RelatedGeneralUtilsTest.java [hunk 11]
```diff
@@ -177,5 +177,35 @@ public void test_getAddressDistrict() {
...
+    assertThat(Hl7RelatedGeneralUtils.getFormattedTelecomNumberValue("","22","333","","x555","112")).isEqualTo("112");
+    // Country, Area, and Extension are used if there is a Local number, and XTN1 and XTN12 are ignored  
+    assertThat(Hl7RelatedGeneralUtils.getFormattedTelecomNumberValue("111","22","333","4444444","x555","112")).isNotEmpty();
+    assertThat(Hl7RelatedGeneralUtils.getFormattedTelecomNumberValue("111","22","333","4444444","x555","112")).contains("+22");
+    assertThat(Hl7RelatedGeneralUtils.getFormattedTelecomNumberValue("111","22","333","4444444","x555","112")).contains("333");
+    assertThat(Hl7RelatedGeneralUtils.getFormattedTelecomNumberValue("111","22","333","4444444","x555","112")).contains("4444");
+    assertThat(Hl7RelatedGeneralUtils.getFormattedTelecomNumberValue("111","22","333","4444444","x555","112")).contains("x555");
+    assertThat(Hl7RelatedGeneralUtils.getFormattedTelecomNumberValue("111","22","333","4444444","x555","112")).isEqualTo("+22 333 444 4444 x555");
+        // Country, Area, and Extension are used if there is a Local number, and XTN1 and XTN12 are ignored
```
[the remark this answers is not in the document]
Comment (cragun47) [reply src/test/java/io/github/linuxforhealth/hl7/data/Hl7RelatedGeneralUtilsTest.java:202]: Fixed.

**Review by LisaWellman — APPROVED** [review]

**Pull request merged**